tests/authenticate_cred_test.cc

```diff
@@ -0,0 +1,63 @@
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include <curl/curl.h>
+#include <boost/process.hpp>
+
+#include "authenticate.h"
+#include "server_credentials.h"
+#include "test_utils.h"
+#include "treehub_server.h"
+#include "utilities/utils.h"
+
+boost::filesystem::path good_test_cred;
+boost::filesystem::path good_auth_json;
+
+/* Authenticate with OAuth2.
+ * Parse authentication information from treehub.json. */
+TEST(authenticate, good_zip) {
+  // Authenticates with the ATS portal to the SaaS instance.
+  ServerCredentials creds(good_test_cred);
+  EXPECT_EQ(creds.GetMethod(), AuthMethod::kOauth2);
+  TreehubServer treehub;
+  int r = authenticate("", creds, treehub);
+  EXPECT_EQ(0, r);
+}
+
+/* Extract credentials from a provided JSON file. */
+TEST(authenticate, good_json) {
+  // Authenticates with the ATS portal to the SaaS instance.
+  // Outdated. we can probably get rid of the whole json-only authentication at this point. T
+  // he last time that was officially supported was over three years ago(2017)
+  // and it's been "deprecated" ever since.
+  TreehubServer treehub;
+  int r = authenticate("", ServerCredentials(good_auth_json), treehub);
+  EXPECT_EQ(0, r);
+}
+
+#ifndef __NO_MAIN__
+int main(int argc, char **argv) {
+  ::testing::InitGoogleTest(&argc, argv);
+  if (argc != 2) {
+    std::cerr << "Error: " << argv[0] << " requires the path to the credential.zip.\n";
+    return EXIT_FAILURE;
+  }
+  good_test_cred = argv[1];
+
+  // prepare auth_test_good.json, use treehub.json from SOTA_PACKED_CREDENTIALS
+  // extract treehub.json from and save it as auth_test_good.json.
+  TemporaryDirectory tmp_data_dir;
+  good_auth_json = tmp_data_dir / "auth_test_good.json";
+
+  int res =
+      system((std::string("unzip -p ") + argv[1] + std::string(" treehub.json >") + good_auth_json.string()).c_str());
```

Comment:
We have `Utils::shell` as a convenience function for system calls.

Reply:
>>I'm not sure I understand why some of the tests were moved to a separate file in a different directory.
there is additional check in cmakelists.txt for test files and suggestion use "test" for everything that require conditional compilation. so there are builds without credentials at all and builds with fake value (dummy_credentials) instead of real file. So real tests happens on gitlab and for other builds there are 2 ways to exclude some test - test dir + if()-endif() and label "credential" on test
https://github.com/advancedtelematic/aktualizr/blob/d5bc8bad0327f3d9bfa474c9a9636f457caed4a3/CMakeLists.txt#L399

Reply:
The correct way to prevent tests that require credentials from running when credentials are not available is to use `if(SOTA_PACKED_CREDENTIALS)` in the CMake code, which is already done for some tests in `src/sota_tools/CMakeLists.txt`.

Using the labels like `credentials` is a good way to control which tests are run if what circumstances, but I think that should be secondary to the if clauses. There is no point in even trying to run tests that require credentials if credentials weren't provided to CMake.